File: src/modules/alac/commands/spec.ts
import { existsSync, unlinkSync } from 'node:fs'
import os from 'node:os'
import path from 'node:path'

import type { Audio, Document, Voice } from '@mtcute/bun'
import { filters } from '@mtcute/dispatcher'

import { generateSpectrogram, probeAudio } from '@/modules/alac/spectrogram.ts'
import { debug, error, infoSpan } from '@/utils/logger.ts'

import type { CommandContext } from './types.ts'
import { parseDynamicHtml } from './types.ts'

const AUDIO_EXTENSIONS = new Set([
  '.m4a',
  '.flac',
  '.mp3',
  '.wav',
  '.wave',
  '.aac',
  '.alac',
  '.ogg',
  '.oga',
  '.opus',
  '.aiff',
  '.aif',
  '.wma',
  '.mka',
  '.ape',
  '.wv',
  '.dsf',
  '.dff',
])

function formatDuration(seconds: number): string {
  const mins = Math.floor(seconds / 60)
  const secs = Math.floor(seconds % 60)
  if (mins >= 60) {
    const hours = Math.floor(mins / 60)
    const remainingMins = mins % 60
    return `${hours}:${String(remainingMins).padStart(2, '0')}:${String(secs).padStart(2, '0')}`
  }
  return `${mins}:${String(secs).padStart(2, '0')}`
}

function escapeHtml(text: string): string {
  return text
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
}

export function registerSpecCommand(ctx: CommandContext): void {
  const { dp, tg, auth } = ctx

  dp.onNewMessage(
    filters.command(['spec', 'spectogram', 'spectrogram', 'spek']),
    async (msg) => {
      using _ = infoSpan('spec_command').enter()

      const isAuthed = await auth.isAuthorized(msg.sender.id, msg.chat.id)
      if (!isAuthed && !auth.isAdmin(msg.sender.id)) {
        debug('Unauthorized user attempted /spec command', {
          user_id: msg.sender.id,
        })
        return
      }

      const reply = await msg.getReplyTo().catch(() => null)
      if (!reply?.media) {
        await msg.replyText(
          parseDynamicHtml(
            '📊 <b>Audio Spectrogram Analyzer</b><br/><br/>' +
              'Reply to any audio file, voice note, or audio document with <code>/spec</code> or <code>/spectogram</code> to generate its frequency spectrogram.<br/><br/>' +
              '<blockquote>💡 <i>Spectrograms visually expose frequency cut-offs (e.g. 16kHz for 128k MP3, 20kHz for 320k MP3), verifying genuine uncompressed lossless masters.</i></blockquote>',
          ),
        )
        return
      }

      const media = reply.media
      let ext = '.audio'
      let isAudio = false
      let fileTitle = ''

      if (media.type === 'audio') {
        isAudio = true
        ext = '.m4a'
        fileTitle = [media.performer, media.title].filter(Boolean).join(' - ')
        const fn = (media as { fileName?: string }).fileName
        if (fn && path.extname(fn)) {
          ext = path.extname(fn).toLowerCase()
        }
      } else if (media.type === 'voice') {
        isAudio = true
        ext = '.ogg'
        fileTitle = 'Voice Note'
      } else if (media.type === 'document') {
        const mime = (media as { mimeType?: string }).mimeType || ''
        const fn =
          (media as { fileName?: string }).fileName ||
          (media as { name?: string }).name ||
          ''
        const fileExt = path.extname(fn).toLowerCase()

        if (mime.startsWith('audio/') || AUDIO_EXTENSIONS.has(fileExt)) {
          isAudio = true
          ext = fileExt || '.m4a'
          fileTitle = fn.replace(/\.[^/.]+$/, '')
        }
      }

      if (
        !isAudio ||
        (media.type !== 'audio' &&
          media.type !== 'voice' &&
          media.type !== 'document')
      ) {
        await msg.replyText(
          parseDynamicHtml(
            '⚠️ <b>Unsupported Media:</b> Please reply to an audio track, voice message, or audio document.',
          ),
        )
        return
      }

      const downloadMedia: Audio | Voice | Document = media

      const statusMsg = await tg.sendText(
        msg.chat.id,
        parseDynamicHtml(
          '⏳ <i>Downloading audio for spectrogram analysis...</i>',
        ),
        { replyTo: msg.id },
      )

      const uniqueId = `${Date.now()}_${Math.random().toString(36).slice(2, 8)}`
      const tmpAudio = path.join(os.tmpdir(), `spec_in_${uniqueId}${ext}`)
      const tmpPng = path.join(os.tmpdir(), `spec_out_${uniqueId}.png`)

      try {
        await tg.downloadToFile(tmpAudio, downloadMedia)

        await tg
          .editMessage({
            chatId: msg.chat.id,
            message: statusMsg.id,
            text: parseDynamicHtml(
              '🔬 <i>Analyzing frequencies &amp; generating spectrogram...</i>',
            ),
          })
          .catch(() => null)

        let probe = null
        try {
          probe = await probeAudio(tmpAudio)
        } catch (probeErr) {
          debug('Audio probe warning (proceeding with spectrogram)', {
            error: String(probeErr),
          })
        }

        const songTitle = probe?.title || fileTitle || 'Audio Track'
        const artist = probe?.artist
        const album = probe?.album
        const codec = (probe?.codec || ext.replace('.', '')).toUpperCase()
        const sampleRate = probe?.sampleRate || 44100
        const bitDepth = probe?.bitDepth
        const bitRate = probe?.bitRate
        const duration = probe?.duration || 0
        const channels = probe?.channels || 2

        const headerTitle = artist ? `${artist} - ${songTitle}` : songTitle
        const headerComment = `${codec}${bitDepth ? ` • ${bitDepth}-bit` : ''} • ${sampleRate.toLocaleString()} Hz${bitRate ? ` • ${Math.round(bitRate / 1000)} kbps` : ''}`

        await generateSpectrogram(tmpAudio, tmpPng, {
          title: headerTitle,
          comment: headerComment,
          duration,
        })

        let caption =
          '📊 <b>Audio Spectrogram Analysis</b><br/><br/>' +
          `• <b>Track:</b> <code>${escapeHtml(songTitle)}</code><br/>`

        if (artist) {
          caption += `• <b>Artist:</b> <code>${escapeHtml(artist)}</code><br/>`
        }
        if (album) {
          caption += `• <b>Album:</b> <code>${escapeHtml(album)}</code><br/>`
        }

        caption +=
          `• <b>Codec:</b> <code>${escapeHtml(codec)}</code><br/>` +
          `• <b>Sample Rate:</b> <code>${sampleRate.toLocaleString()} Hz (${Math.round(sampleRate / 100) / 10} kHz)</code><br/>`

        if (bitDepth) {
          caption += `• <b>Bit Depth:</b> <code>${bitDepth}-bit</code><br/>`
        }

        caption += `• <b>Channels:</b> <code>${channels === 2 ? 'Stereo (2.0)' : channels === 1 ? 'Mono' : `${channels} channels`}</code><br/>`

        if (bitRate) {
          caption += `• <b>Bitrate:</b> <code>${Math.round(bitRate / 1000)} kbps</code><br/>`
        }

        if (duration > 0) {
          caption += `• <b>Duration:</b> <code>${formatDuration(duration)}</code><br/>`
        }

        await tg.sendMedia(
          msg.chat.id,
          {
            type: 'photo',
            file: Bun.file(tmpPng),
            caption: parseDynamicHtml(caption),
          },
          { replyTo: reply.id },
        )

        await tg
          .deleteMessagesById(msg.chat.id, [statusMsg.id])
          .catch(() => null)
      } catch (err) {
        error('Failed to generate spectrogram', { error: String(err) })
        await tg
          .editMessage({
            chatId: msg.chat.id,
            message: statusMsg.id,
            text: parseDynamicHtml(
              `❌ <b>Spectrogram Generation Failed:</b> <code>${escapeHtml(String(err))}</code>`,
            ),
          })
          .catch(() => null)
      } finally {
        if (existsSync(tmpAudio)) {
          try {
            unlinkSync(tmpAudio)
          } catch {}
        }
        if (existsSync(tmpPng)) {
          try {
            unlinkSync(tmpPng)
          } catch {}
        }
      }
    },
  )
}

File: crates/bot/src/handlers/spec.rs
use std::{
    path::{Path, PathBuf},
    sync::{
        atomic::{AtomicU64, Ordering},
        Arc,
    },
    time::{SystemTime, UNIX_EPOCH},
};

use ferogram::{filters, filters::Dispatcher, InputMessage, PeerRef};

use crate::{
    html::{escape, parse_dynamic_html},
    spectrogram::{self, AudioProbeResult, AUDIO_EXTENSIONS},
    BotState,
};

const USAGE: &str = "📊 <b>Audio Spectrogram Analyzer</b><br/><br/>Reply to any audio file, voice note, or audio document with <code>/spec</code> or <code>/spectogram</code> to generate its frequency spectrogram.<br/><br/><blockquote>💡 <i>Spectrograms visually expose frequency cut-offs (e.g. 16kHz for 128k MP3, 20kHz for 320k MP3), verifying genuine uncompressed lossless masters.</i></blockquote>";
const UNSUPPORTED: &str = "⚠️ <b>Unsupported Media:</b> Please reply to an audio track, voice message, or audio document.";
static NEXT_TEMP_ID: AtomicU64 = AtomicU64::new(0);

struct ClassifiedMedia {
    media: ferogram::tl::enums::MessageMedia,
    ext: String,
    file_title: String,
}

fn classify_media(media: &ferogram::tl::enums::MessageMedia) -> Option<ClassifiedMedia> {
    let ferogram::tl::enums::MessageMedia::Document(_) = media else {
        return None;
    };
    let document = ferogram::media::Document::from_media(media)?;
    let is_voice = document.raw.attributes.iter().any(|attribute| {
        matches!(attribute, ferogram::tl::enums::DocumentAttribute::Audio(audio) if audio.voice)
    });
    let filename = document.file_name().unwrap_or_default();
    let file_ext = Path::new(filename)
        .extension()
        .map(|ext| format!(".{}", ext.to_string_lossy().to_lowercase()));
    let mime = document.mime_type();
    let is_audio = is_voice
        || mime.starts_with("audio/")
        || file_ext
            .as_deref()
            .is_some_and(|ext| AUDIO_EXTENSIONS.contains(&ext));
    if !is_audio {
        return None;
    }
    let (ext, file_title) = if is_voice {
        (".ogg".to_owned(), "Voice Note".to_owned())
    } else {
        let ext = file_ext.unwrap_or_else(|| ".m4a".to_owned());
        let file_title = filename
            .rfind('.')
            .map_or_else(|| filename.to_owned(), |index| filename[..index].to_owned());
        (ext, file_title)
    };
    Some(ClassifiedMedia {
        media: media.clone(),
        ext,
        file_title,
    })
}

fn temp_paths(ext: &str) -> (PathBuf, PathBuf) {
    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_millis())
        .unwrap_or_default();
    let unique = NEXT_TEMP_ID.fetch_add(1, Ordering::Relaxed);
    let id = format!("{timestamp}_{unique}");
    (
        PathBuf::from(format!("/tmp/opencode/spec_in_{id}{ext}")),
        PathBuf::from(format!("/tmp/opencode/spec_out_{id}.png")),
    )
}

async fn edit_status(state: &BotState, peer: &PeerRef, id: i32, text: &str) {
    let _ = state
        .client
        .edit_message(
            peer.clone(),
            id,
            InputMessage::html(parse_dynamic_html(text)),
        )
        .await;
}

async fn delete_status(state: &BotState, peer: &PeerRef, id: i32) {
    if let Ok(messages) = state.client.get_messages(peer.clone(), &[id]).await {
        if let Some(message) = messages.first() {
            let _ = message.delete().await;
        }
    }
}

async fn handle(state: Arc<BotState>, msg: ferogram::update::IncomingMessage) {
    let sender = msg.sender_user_id().unwrap_or_default();
    let chat = super::marked_chat_id(&msg);
    let authorized = state
        .auth
        .is_authorized(sender, Some(chat))
        .await
        .unwrap_or(false);
    if !authorized && !state.auth.is_admin(sender) {
        return;
    }

    let peer = super::chat_peer_ref(&msg);
    let Some(reply_id) = msg.reply_to_message_id() else {
        let _ = msg
            .reply(InputMessage::html(parse_dynamic_html(USAGE)))
            .await;
        return;
    };
    let Ok(replies) = state.client.get_messages(peer.clone(), &[reply_id]).await else {
        let _ = msg
            .reply(InputMessage::html(parse_dynamic_html(USAGE)))
            .await;
        return;
    };
    let Some(reply) = replies.first() else {
        let _ = msg
            .reply(InputMessage::html(parse_dynamic_html(USAGE)))
            .await;
        return;
    };
    let Some(media) = reply.media().and_then(classify_media) else {
        let text = if reply.media().is_some() {
            UNSUPPORTED
        } else {
            USAGE
        };
        let _ = msg
            .reply(InputMessage::html(parse_dynamic_html(text)))
            .await;
        return;
    };

    let Ok(status) = state
        .client
        .send_message(
            peer.clone(),
            InputMessage::html(parse_dynamic_html(
                "⏳ <i>Downloading audio for spectrogram analysis...</i>",
            ))
            .reply_to(Some(msg.id())),
        )
        .await
    else {
        return;
    };
    let (input_path, output_path) = temp_paths(&media.ext);
    let result = async {
        state
            .client
            .download_media(
                &media.media,
                ferogram::media::MediaQuality::Original,
                &input_path,
                None,
            )
            .await
            .map_err(|error| error.to_string())?;
        edit_status(
            &state,
            &peer,
            status.id(),
            "🔬 <i>Analyzing frequencies &amp; generating spectrogram...</i>",
        )
        .await;

        let probe = spectrogram::probe_audio(&input_path)
            .await
            .unwrap_or_else(|_| AudioProbeResult {
                title: None,
                artist: None,
                album: None,
                codec: media.ext.trim_start_matches('.').to_owned(),
                sample_rate: 44_100,
                bit_depth: None,
                channels: 2,
                bit_rate: None,
                duration: 0.0,
            });
        let song_title = probe.title.clone().unwrap_or_else(|| {
            if media.file_title.is_empty() {
                "Audio Track".to_owned()
            } else {
                media.file_title.clone()
            }
        });
        let codec = probe.codec.to_uppercase();
        let header_title = probe.artist.as_ref().map_or_else(
            || song_title.clone(),
            |artist| format!("{artist} - {song_title}"),
        );
        let header_comment = format!(
            "{}{} • {} Hz{}",
            codec,
            probe
                .bit_depth
                .map_or_else(String::new, |depth| format!(" • {depth}-bit")),
            spectrogram::thousands_separator(probe.sample_rate),
            probe.bit_rate.map_or_else(String::new, |rate| format!(
                " • {} kbps",
                (rate as f64 / 1000.0).round()
            ))
        );
        spectrogram::generate_spectrogram(
            &input_path,
            &output_path,
            Some(&header_title),
            Some(&header_comment),
            probe.duration,
        )
        .await?;
        let uploaded = state
            .client
            .upload_file(&output_path)
            .await
            .map_err(|error| error.to_string())?;
        let caption = spectrogram::build_caption(&probe, &song_title, &codec);
        state
            .client
            .send_message(
                peer.clone(),
                InputMessage::html(parse_dynamic_html(&caption))
                    .copy_media(uploaded.as_photo_media())
                    .reply_to(Some(reply.id())),
            )
            .await
            .map_err(|error| error.to_string())?;
        delete_status(&state, &peer, status.id()).await;
        Ok::<(), String>(())
    }
    .await;
    if let Err(error) = result {
        edit_status(
            &state,
            &peer,
            status.id(),
            &format!(
                "❌ <b>Spectrogram Generation Failed:</b> <code>{}</code>",
                escape(&error)
            ),
        )
        .await;
    }
    let _ = tokio::fs::remove_file(input_path).await;
    let _ = tokio::fs::remove_file(output_path).await;
}

pub fn register(dp: &mut Dispatcher, state: Arc<BotState>) {
    for command in ["spec", "spectogram", "spectrogram", "spek"] {
        let state = Arc::clone(&state);
        dp.on_message(filters::command(command), move |msg| {
            let state = Arc::clone(&state);
            async move { handle(state, msg).await }
        });
    }
}
